SquirrelPanel.m

```diff
@@ -975,31 +975,41 @@ - (void)showPreedit:(NSString *)preedit
     if (labelRange.location == NSNotFound) {
       labelRange2 = labelRange;
       labelFormat2 = labelFormat = nil;
-      pureCandidateRange = NSMakeRange(0, candidateFormat.length);
-    } else {
       pureCandidateRange = [candidateFormat rangeOfString:@"%@"];
       if (pureCandidateRange.location == NSNotFound) {
-        // this should never happen, but just ensure that Squirrel
-        // would not crash when such edge case occurs...
-        labelFormat = candidateFormat;
-        labelRange2 = pureCandidateRange;
+        pureCandidateRange = NSMakeRange(0, candidateFormat.length);
+      }
+    }
+    pureCandidateRange = [candidateFormat rangeOfString:@"%@"];
+    if (pureCandidateRange.location == NSNotFound) {
+      // this should never happen, but just ensure that Squirrel
+      // would not crash when such edge case occurs...
+      labelFormat = candidateFormat;
+      labelRange2 = pureCandidateRange;
+      labelFormat2 = nil;
+    } else {
+      if (NSMaxRange(pureCandidateRange) >= candidateFormat.length) {
+        // '%@' is at the end, so label2 does not exist
+        labelRange2 = NSMakeRange(NSNotFound, 0);
         labelFormat2 = nil;
+        // fix label1, everything other than '%@' is label1
+        labelRange.location = 0;
+        labelRange.length = pureCandidateRange.location;
       } else {
-        if (NSMaxRange(pureCandidateRange) >= candidateFormat.length) {
-          // '%@' is at the end, so label2 does not exist
-          labelRange2 = NSMakeRange(NSNotFound, 0);
-          labelFormat2 = nil;
-          // fix label1, everything other than '%@' is label1
-          labelRange.location = 0;
-          labelRange.length = pureCandidateRange.location;
-        } else {
+        if (pureCandidateRange.location > 0) {
           labelRange = NSMakeRange(0, pureCandidateRange.location);
-          labelRange2 = NSMakeRange(NSMaxRange(pureCandidateRange),
-                                    candidateFormat.length -
-                                        NSMaxRange(pureCandidateRange));
-          labelFormat2 = [candidateFormat substringWithRange:labelRange2];
+        } else {
+          labelRange = NSMakeRange(NSNotFound, 0);
         }
+        labelRange2 = NSMakeRange(NSMaxRange(pureCandidateRange),
+                                  candidateFormat.length -
+                                      NSMaxRange(pureCandidateRange));
+        labelFormat2 = [candidateFormat substringWithRange:labelRange2];
+      }
+      if (labelRange.location != NSNotFound) {
         labelFormat = [candidateFormat substringWithRange:labelRange];
+      } else {
+        labelFormat = nil;
```

Comment:
In the latest version,
`labelRange.location` can be `NSNotFound`, set `labelFormat` to `nil`, since `labelFormat` is the only exposed property of `theme`.
`labelRange` is a local variable only.

Reply:
Wow I mean all the above change, not just this line.
There's too many nested if-else so I need more time to follow up the changes happening in it.
I believe the logic can be made simpler.